{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66053f89",
   "metadata": {},
   "source": [
    "## Frameshift finder\n",
    "### Import fasta file\n",
    "### Translate file need to use a single translation table. Need to group fasta input by tax groups if using alternative translation tables\n",
    "### Create initial alignments of DNA and protein sequences\n",
    "### Scan the DNA alignment for gaps < 3 bp. If gap is found, these are reported along with the location. The gap then converted N, the sequence is retranslated and relaligned to determine if the read frame can be restored. \n",
    "### Need to consider insertion frameshifts as well as deletions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4580aa87",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import pandas as pd \n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c239028d",
   "metadata": {},
   "source": [
    "### Read input file and translate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "1f208706",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1 enter files names and variables\n",
    "nucleotide = \"newcytb.fsa\"\n",
    "protein_trans = \"newcytb.pro\"\n",
    "nuc_alignment = \"newcytb.aln\"\n",
    "prot_alignment = \"newcytb_pro.aln\"\n",
    "nuc_alignment_fsa = \"newcytb_pro.fsaaln\"\n",
    "\n",
    "translation_table = \"2\"\n",
    "\n",
    "revised_nucs = \"revisedfasta\"\n",
    "revised_nucs_aln = \"revisedfasta.aln\"\n",
    "revised_nucs_alnfsa = \"revisedfasta.fsaaln\"\n",
    "revised_pro = \"revisedprotein\"\n",
    "revised_pro_aln = \"revisedprotein.aln\"\n",
    "revised_pro_alnfsa = \"revisedprotein.fsaaln\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5bb2999e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 2\n",
    "from Bio import SeqIO\n",
    "#coding_dna = []\n",
    "protein_seq = []\n",
    "#trans = []\n",
    "#translations = []\n",
    "for seq_record in SeqIO.parse(nucleotide, \"fasta\"):\n",
    "    s = seq_record\n",
    "    #coding_dna = s.seq\n",
    "    s.seq = s.seq.translate(table = translation_table, to_stop=True)\n",
    "    protein_seq.append(s)\n",
    "#print(protein_seq)\n",
    "SeqIO.write(protein_seq, protein_trans, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79e0084a",
   "metadata": {},
   "source": [
    "### Create alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "ba375628",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.AlignIO in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.AlignIO - Multiple sequence alignment input/output as alignment objects.\n",
      "\n",
      "DESCRIPTION\n",
      "    The Bio.AlignIO interface is deliberately very similar to Bio.SeqIO, and in\n",
      "    fact the two are connected internally.  Both modules use the same set of file\n",
      "    format names (lower case strings).  From the user's perspective, you can read\n",
      "    in a PHYLIP file containing one or more alignments using Bio.AlignIO, or you\n",
      "    can read in the sequences within these alignments using Bio.SeqIO.\n",
      "    \n",
      "    Bio.AlignIO is also documented at http://biopython.org/wiki/AlignIO and by\n",
      "    a whole chapter in our tutorial:\n",
      "    \n",
      "    * `HTML Tutorial`_\n",
      "    * `PDF Tutorial`_\n",
      "    \n",
      "    .. _`HTML Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
      "    .. _`PDF Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.pdf\n",
      "    \n",
      "    Input\n",
      "    -----\n",
      "    For the typical special case when your file or handle contains one and only\n",
      "    one alignment, use the function Bio.AlignIO.read().  This takes an input file\n",
      "    handle (or in recent versions of Biopython a filename as a string), format\n",
      "    string and optional number of sequences per alignment.  It will return a single\n",
      "    MultipleSeqAlignment object (or raise an exception if there isn't just one\n",
      "    alignment):\n",
      "    \n",
      "    >>> from Bio import AlignIO\n",
      "    >>> align = AlignIO.read(\"Phylip/interlaced.phy\", \"phylip\")\n",
      "    >>> print(align)\n",
      "    Alignment with 3 rows and 384 columns\n",
      "    -----MKVILLFVLAVFTVFVSS---------------RGIPPE...I-- CYS1_DICDI\n",
      "    MAHARVLLLALAVLATAAVAVASSSSFADSNPIRPVTDRAASTL...VAA ALEU_HORVU\n",
      "    ------MWATLPLLCAGAWLLGV--------PVCGAAELSVNSL...PLV CATH_HUMAN\n",
      "    \n",
      "    For the general case, when the handle could contain any number of alignments,\n",
      "    use the function Bio.AlignIO.parse(...) which takes the same arguments, but\n",
      "    returns an iterator giving MultipleSeqAlignment objects (typically used in a\n",
      "    for loop). If you want random access to the alignments by number, turn this\n",
      "    into a list:\n",
      "    \n",
      "    >>> from Bio import AlignIO\n",
      "    >>> alignments = list(AlignIO.parse(\"Emboss/needle.txt\", \"emboss\"))\n",
      "    >>> print(alignments[2])\n",
      "    Alignment with 2 rows and 120 columns\n",
      "    -KILIVDDQYGIRILLNEVFNKEGYQTFQAANGLQALDIVTKER...--- ref_rec\n",
      "    LHIVVVDDDPGTCVYIESVFAELGHTCKSFVRPEAAEEYILTHP...HKE gi|94967506|receiver\n",
      "    \n",
      "    Most alignment file formats can be concatenated so as to hold as many\n",
      "    different multiple sequence alignments as possible.  One common example\n",
      "    is the output of the tool seqboot in the PHLYIP suite.  Sometimes there\n",
      "    can be a file header and footer, as seen in the EMBOSS alignment output.\n",
      "    \n",
      "    Output\n",
      "    ------\n",
      "    Use the function Bio.AlignIO.write(...), which takes a complete set of\n",
      "    Alignment objects (either as a list, or an iterator), an output file handle\n",
      "    (or filename in recent versions of Biopython) and of course the file format::\n",
      "    \n",
      "        from Bio import AlignIO\n",
      "        alignments = ...\n",
      "        count = SeqIO.write(alignments, \"example.faa\", \"fasta\")\n",
      "    \n",
      "    If using a handle make sure to close it to flush the data to the disk::\n",
      "    \n",
      "        from Bio import AlignIO\n",
      "        alignments = ...\n",
      "        with open(\"example.faa\", \"w\") as handle:\n",
      "            count = SeqIO.write(alignments, handle, \"fasta\")\n",
      "    \n",
      "    In general, you are expected to call this function once (with all your\n",
      "    alignments) and then close the file handle.  However, for file formats\n",
      "    like PHYLIP where multiple alignments are stored sequentially (with no file\n",
      "    header and footer), then multiple calls to the write function should work as\n",
      "    expected when using handles.\n",
      "    \n",
      "    If you are using a filename, the repeated calls to the write functions will\n",
      "    overwrite the existing file each time.\n",
      "    \n",
      "    Conversion\n",
      "    ----------\n",
      "    The Bio.AlignIO.convert(...) function allows an easy interface for simple\n",
      "    alignment file format conversions. Additionally, it may use file format\n",
      "    specific optimisations so this should be the fastest way too.\n",
      "    \n",
      "    In general however, you can combine the Bio.AlignIO.parse(...) function with\n",
      "    the Bio.AlignIO.write(...) function for sequence file conversion. Using\n",
      "    generator expressions provides a memory efficient way to perform filtering or\n",
      "    other extra operations as part of the process.\n",
      "    \n",
      "    File Formats\n",
      "    ------------\n",
      "    When specifying the file format, use lowercase strings.  The same format\n",
      "    names are also used in Bio.SeqIO and include the following:\n",
      "    \n",
      "      - clustal -   Output from Clustal W or X.\n",
      "      - emboss    - EMBOSS tools' \"pairs\" and \"simple\" alignment formats.\n",
      "      - fasta     - The generic sequence file format where each record starts with\n",
      "        an identifier line starting with a \">\" character, followed by\n",
      "        lines of sequence.\n",
      "      - fasta-m10 - For the pairwise alignments output by Bill Pearson's FASTA\n",
      "        tools when used with the -m 10 command line option for machine\n",
      "        readable output.\n",
      "      - ig        - The IntelliGenetics file format, apparently the same as the\n",
      "        MASE alignment format.\n",
      "      - msf       - The GCG MSF alignment format, originally from PileUp tool.\n",
      "      - nexus     - Output from NEXUS, see also the module Bio.Nexus which can also\n",
      "        read any phylogenetic trees in these files.\n",
      "      - phylip    - Interlaced PHYLIP, as used by the PHYLIP tools.\n",
      "      - phylip-sequential - Sequential PHYLIP.\n",
      "      - phylip-relaxed - PHYLIP like format allowing longer names.\n",
      "      - stockholm - A richly annotated alignment file format used by PFAM.\n",
      "      - mauve - Output from progressiveMauve/Mauve\n",
      "    \n",
      "    Note that while Bio.AlignIO can read all the above file formats, it cannot\n",
      "    write to all of them.\n",
      "    \n",
      "    You can also use any file format supported by Bio.SeqIO, such as \"fasta\" or\n",
      "    \"ig\" (which are listed above), PROVIDED the sequences in your file are all the\n",
      "    same length.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    ClustalIO\n",
      "    EmbossIO\n",
      "    FastaIO\n",
      "    Interfaces\n",
      "    MafIO\n",
      "    MauveIO\n",
      "    MsfIO\n",
      "    NexusIO\n",
      "    PhylipIO\n",
      "    StockholmIO\n",
      "\n",
      "FUNCTIONS\n",
      "    convert(in_file, in_format, out_file, out_format, molecule_type=None)\n",
      "        Convert between two alignment files, returns number of alignments.\n",
      "        \n",
      "        Arguments:\n",
      "         - in_file - an input handle or filename\n",
      "         - in_format - input file format, lower case string\n",
      "         - output - an output handle or filename\n",
      "         - out_file - output file format, lower case string\n",
      "         - molecule_type - optional molecule type to apply, string containing\n",
      "           \"DNA\", \"RNA\" or \"protein\".\n",
      "        \n",
      "        **NOTE** - If you provide an output filename, it will be opened which will\n",
      "        overwrite any existing file without warning. This may happen if even the\n",
      "        conversion is aborted (e.g. an invalid out_format name is given).\n",
      "        \n",
      "        Some output formats require the molecule type be specified where this\n",
      "        cannot be determined by the parser. For example, converting to FASTA,\n",
      "        Clustal, or PHYLIP format to NEXUS:\n",
      "        \n",
      "        >>> from io import StringIO\n",
      "        >>> from Bio import AlignIO\n",
      "        >>> handle = StringIO()\n",
      "        >>> AlignIO.convert(\"Phylip/horses.phy\", \"phylip\", handle, \"nexus\", \"DNA\")\n",
      "        1\n",
      "        >>> print(handle.getvalue())\n",
      "        #NEXUS\n",
      "        begin data;\n",
      "        dimensions ntax=10 nchar=40;\n",
      "        format datatype=dna missing=? gap=-;\n",
      "        matrix\n",
      "        Mesohippus   AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
      "        Hypohippus   AAACCCCCCCAAAAAAAAACAAAAAAAAAAAAAAAAAAAA\n",
      "        Archaeohip   CAAAAAAAAAAAAAAAACACAAAAAAAAAAAAAAAAAAAA\n",
      "        Parahippus   CAAACAACAACAAAAAAAACAAAAAAAAAAAAAAAAAAAA\n",
      "        Merychippu   CCAACCACCACCCCACACCCAAAAAAAAAAAAAAAAAAAA\n",
      "        'M. secundu' CCAACCACCACCCACACCCCAAAAAAAAAAAAAAAAAAAA\n",
      "        Nannipus     CCAACCACAACCCCACACCCAAAAAAAAAAAAAAAAAAAA\n",
      "        Neohippari   CCAACCCCCCCCCCACACCCAAAAAAAAAAAAAAAAAAAA\n",
      "        Calippus     CCAACCACAACCCACACCCCAAAAAAAAAAAAAAAAAAAA\n",
      "        Pliohippus   CCCACCCCCCCCCACACCCCAAAAAAAAAAAAAAAAAAAA\n",
      "        ;\n",
      "        end;\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    parse(handle, format, seq_count=None)\n",
      "        Iterate over an alignment file as MultipleSeqAlignment objects.\n",
      "        \n",
      "        Arguments:\n",
      "         - handle    - handle to the file, or the filename as a string\n",
      "           (note older versions of Biopython only took a handle).\n",
      "         - format    - string describing the file format.\n",
      "         - seq_count - Optional integer, number of sequences expected in each\n",
      "           alignment.  Recommended for fasta format files.\n",
      "        \n",
      "        If you have the file name in a string 'filename', use:\n",
      "        \n",
      "        >>> from Bio import AlignIO\n",
      "        >>> filename = \"Emboss/needle.txt\"\n",
      "        >>> format = \"emboss\"\n",
      "        >>> for alignment in AlignIO.parse(filename, format):\n",
      "        ...     print(\"Alignment of length %i\" % alignment.get_alignment_length())\n",
      "        Alignment of length 124\n",
      "        Alignment of length 119\n",
      "        Alignment of length 120\n",
      "        Alignment of length 118\n",
      "        Alignment of length 125\n",
      "        \n",
      "        If you have a string 'data' containing the file contents, use::\n",
      "        \n",
      "          from Bio import AlignIO\n",
      "          from io import StringIO\n",
      "          my_iterator = AlignIO.parse(StringIO(data), format)\n",
      "        \n",
      "        Use the Bio.AlignIO.read() function when you expect a single record only.\n",
      "    \n",
      "    read(handle, format, seq_count=None)\n",
      "        Turn an alignment file into a single MultipleSeqAlignment object.\n",
      "        \n",
      "        Arguments:\n",
      "         - handle    - handle to the file, or the filename as a string\n",
      "           (note older versions of Biopython only took a handle).\n",
      "         - format    - string describing the file format.\n",
      "         - seq_count - Optional integer, number of sequences expected in each\n",
      "           alignment.  Recommended for fasta format files.\n",
      "        \n",
      "        If the handle contains no alignments, or more than one alignment,\n",
      "        an exception is raised.  For example, using a PFAM/Stockholm file\n",
      "        containing one alignment:\n",
      "        \n",
      "        >>> from Bio import AlignIO\n",
      "        >>> filename = \"Clustalw/protein.aln\"\n",
      "        >>> format = \"clustal\"\n",
      "        >>> alignment = AlignIO.read(filename, format)\n",
      "        >>> print(\"Alignment of length %i\" % alignment.get_alignment_length())\n",
      "        Alignment of length 411\n",
      "        \n",
      "        If however you want the first alignment from a file containing\n",
      "        multiple alignments this function would raise an exception.\n",
      "        \n",
      "        >>> from Bio import AlignIO\n",
      "        >>> filename = \"Emboss/needle.txt\"\n",
      "        >>> format = \"emboss\"\n",
      "        >>> alignment = AlignIO.read(filename, format)\n",
      "        Traceback (most recent call last):\n",
      "            ...\n",
      "        ValueError: More than one record found in handle\n",
      "        \n",
      "        Instead use:\n",
      "        \n",
      "        >>> from Bio import AlignIO\n",
      "        >>> filename = \"Emboss/needle.txt\"\n",
      "        >>> format = \"emboss\"\n",
      "        >>> alignment = next(AlignIO.parse(filename, format))\n",
      "        >>> print(\"First alignment has length %i\" % alignment.get_alignment_length())\n",
      "        First alignment has length 124\n",
      "        \n",
      "        You must use the Bio.AlignIO.parse() function if you want to read multiple\n",
      "        records from the handle.\n",
      "    \n",
      "    write(alignments, handle, format)\n",
      "        Write complete set of alignments to a file.\n",
      "        \n",
      "        Arguments:\n",
      "         - alignments - A list (or iterator) of MultipleSeqAlignment objects,\n",
      "           or a single alignment object.\n",
      "         - handle    - File handle object to write to, or filename as string\n",
      "           (note older versions of Biopython only took a handle).\n",
      "         - format    - lower case string describing the file format to write.\n",
      "        \n",
      "        You should close the handle after calling this function.\n",
      "        \n",
      "        Returns the number of alignments written (as an integer).\n",
      "\n",
      "FILE\n",
      "    /Users/rich/opt/anaconda3/lib/python3.9/site-packages/Bio/AlignIO/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "help(AlignIO)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35fd4db2",
   "metadata": {},
   "source": [
    "### Option to read an alignment prepared with another tool and saved locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "213112eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from Bio import AlignIO\n",
    "\n",
    "#alignment = AlignIO.read(open(\"align.out\"), \"fasta\")\n",
    "#print(\"Alignment length %i\" % alignment.get_alignment_length())\n",
    "#for record in alignment:\n",
    "#    print(record.seq + \" \" + record.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f96bad8",
   "metadata": {},
   "source": [
    "## Create clustal alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "32c8939d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class ClustalwCommandline in module Bio.Align.Applications._Clustalw:\n",
      "\n",
      "class ClustalwCommandline(Bio.Application.AbstractCommandline)\n",
      " |  ClustalwCommandline(cmd='clustalw', **kwargs)\n",
      " |  \n",
      " |  Command line wrapper for clustalw (version one or two).\n",
      " |  \n",
      " |  http://www.clustal.org/\n",
      " |  \n",
      " |  Notes\n",
      " |  -----\n",
      " |  Last checked against versions: 1.83 and 2.1\n",
      " |  \n",
      " |  References\n",
      " |  ----------\n",
      " |  Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA,\n",
      " |  McWilliam H, Valentin F, Wallace IM, Wilm A, Lopez R, Thompson JD,\n",
      " |  Gibson TJ, Higgins DG. (2007). Clustal W and Clustal X version 2.0.\n",
      " |  Bioinformatics, 23, 2947-2948.\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> from Bio.Align.Applications import ClustalwCommandline\n",
      " |  >>> in_file = \"unaligned.fasta\"\n",
      " |  >>> clustalw_cline = ClustalwCommandline(\"clustalw2\", infile=in_file)\n",
      " |  >>> print(clustalw_cline)\n",
      " |  clustalw2 -infile=unaligned.fasta\n",
      " |  \n",
      " |  You would typically run the command line with clustalw_cline() or via\n",
      " |  the Python subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      ClustalwCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='clustalw', **kwargs)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  align\n",
      " |      Do full multiple alignment.\n",
      " |      \n",
      " |      This property controls the addition of the -align switch, treat this property as a boolean.\n",
      " |  \n",
      " |  bootlabels\n",
      " |      Node OR branch position of bootstrap values in tree display\n",
      " |      \n",
      " |      This controls the addition of the -bootlabels parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  bootstrap\n",
      " |      Bootstrap a NJ tree (n= number of bootstraps; def. = 1000).\n",
      " |      \n",
      " |      This controls the addition of the -bootstrap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  case\n",
      " |      LOWER or UPPER (for GDE output only)\n",
      " |      \n",
      " |      This controls the addition of the -case parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  check\n",
      " |      Outline the command line params.\n",
      " |      \n",
      " |      This property controls the addition of the -check switch, treat this property as a boolean.\n",
      " |  \n",
      " |  clustering\n",
      " |      NJ or UPGMA\n",
      " |      \n",
      " |      This controls the addition of the -clustering parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  convert\n",
      " |      Output the input sequences in a different file format.\n",
      " |      \n",
      " |      This property controls the addition of the -convert switch, treat this property as a boolean.\n",
      " |  \n",
      " |  dnamatrix\n",
      " |      DNA weight matrix=IUB, CLUSTALW or filename\n",
      " |      \n",
      " |      This controls the addition of the -dnamatrix parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  endgaps\n",
      " |      No end gap separation pen.\n",
      " |      \n",
      " |      This property controls the addition of the -endgaps switch, treat this property as a boolean.\n",
      " |  \n",
      " |  fullhelp\n",
      " |      Output full help content.\n",
      " |      \n",
      " |      This property controls the addition of the -fullhelp switch, treat this property as a boolean.\n",
      " |  \n",
      " |  gapdist\n",
      " |      Gap separation pen. range\n",
      " |      \n",
      " |      This controls the addition of the -gapdist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapext\n",
      " |      Gap extension penalty\n",
      " |      \n",
      " |      This controls the addition of the -gapext parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapopen\n",
      " |      Gap opening penalty\n",
      " |      \n",
      " |      This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  helixendin\n",
      " |      Number of residues inside helix to be treated as terminal\n",
      " |      \n",
      " |      This controls the addition of the -helixendin parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  helixendout\n",
      " |      Number of residues outside helix to be treated as terminal\n",
      " |      \n",
      " |      This controls the addition of the -helixendout parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  helixgap\n",
      " |      Gap penalty for helix core residues\n",
      " |      \n",
      " |      This controls the addition of the -helixgap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  help\n",
      " |      Outline the command line params.\n",
      " |      \n",
      " |      This property controls the addition of the -help switch, treat this property as a boolean.\n",
      " |  \n",
      " |  hgapresidues\n",
      " |      List hydrophilic res.\n",
      " |      \n",
      " |      This property controls the addition of the -hgapresidues switch, treat this property as a boolean.\n",
      " |  \n",
      " |  infile\n",
      " |      Input sequences.\n",
      " |      \n",
      " |      This controls the addition of the -infile parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  iteration\n",
      " |      NONE or TREE or ALIGNMENT\n",
      " |      \n",
      " |      This controls the addition of the -iteration parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  kimura\n",
      " |      Use Kimura's correction.\n",
      " |      \n",
      " |      This property controls the addition of the -kimura switch, treat this property as a boolean.\n",
      " |  \n",
      " |  ktuple\n",
      " |      Word size\n",
      " |      \n",
      " |      This controls the addition of the -ktuple parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  loopgap\n",
      " |      Gap penalty for loop regions\n",
      " |      \n",
      " |      This controls the addition of the -loopgap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  matrix\n",
      " |      Protein weight matrix=BLOSUM, PAM, GONNET, ID or filename\n",
      " |      \n",
      " |      This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  maxdiv\n",
      " |      % ident. for delay\n",
      " |      \n",
      " |      This controls the addition of the -maxdiv parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  maxseqlen\n",
      " |      Maximum allowed input sequence length\n",
      " |      \n",
      " |      This controls the addition of the -maxseqlen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  negative\n",
      " |      Protein alignment with negative values in matrix\n",
      " |      \n",
      " |      This property controls the addition of the -negative switch, treat this property as a boolean.\n",
      " |  \n",
      " |  newtree\n",
      " |      Output file name for newly created guide tree\n",
      " |      \n",
      " |      This controls the addition of the -newtree parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  newtree1\n",
      " |      Output file name for new guide tree of profile1\n",
      " |      \n",
      " |      This controls the addition of the -newtree1 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  newtree2\n",
      " |      Output file for new guide tree of profile2\n",
      " |      \n",
      " |      This controls the addition of the -newtree2 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  nohgap\n",
      " |      Hydrophilic gaps off\n",
      " |      \n",
      " |      This property controls the addition of the -nohgap switch, treat this property as a boolean.\n",
      " |  \n",
      " |  nopgap\n",
      " |      Residue-specific gaps off\n",
      " |      \n",
      " |      This property controls the addition of the -nopgap switch, treat this property as a boolean.\n",
      " |  \n",
      " |  nosecstr1\n",
      " |      Do not use secondary structure-gap penalty mask for profile 1\n",
      " |      \n",
      " |      This property controls the addition of the -nosecstr1 switch, treat this property as a boolean.\n",
      " |  \n",
      " |  nosecstr2\n",
      " |      Do not use secondary structure-gap penalty mask for profile 2\n",
      " |      \n",
      " |      This property controls the addition of the -nosecstr2 switch, treat this property as a boolean.\n",
      " |  \n",
      " |  noweights\n",
      " |      Disable sequence weighting\n",
      " |      \n",
      " |      This property controls the addition of the -noweights switch, treat this property as a boolean.\n",
      " |  \n",
      " |  numiter\n",
      " |      maximum number of iterations to perform\n",
      " |      \n",
      " |      This controls the addition of the -numiter parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  options\n",
      " |      List the command line parameters\n",
      " |      \n",
      " |      This property controls the addition of the -options switch, treat this property as a boolean.\n",
      " |  \n",
      " |  outfile\n",
      " |      Output sequence alignment file name\n",
      " |      \n",
      " |      This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outorder\n",
      " |      Output taxon order: INPUT or ALIGNED\n",
      " |      \n",
      " |      This controls the addition of the -outorder parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  output\n",
      " |      Output format: CLUSTAL(default), GCG, GDE, PHYLIP, PIR, NEXUS and FASTA\n",
      " |      \n",
      " |      This controls the addition of the -output parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outputtree\n",
      " |      nj OR phylip OR dist OR nexus\n",
      " |      \n",
      " |      This controls the addition of the -outputtree parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pairgap\n",
      " |      Gap penalty\n",
      " |      \n",
      " |      This controls the addition of the -pairgap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pim\n",
      " |      Output percent identity matrix (while calculating the tree).\n",
      " |      \n",
      " |      This property controls the addition of the -pim switch, treat this property as a boolean.\n",
      " |  \n",
      " |  profile\n",
      " |      Merge two alignments by profile alignment\n",
      " |      \n",
      " |      This property controls the addition of the -profile switch, treat this property as a boolean.\n",
      " |  \n",
      " |  profile1\n",
      " |      Profiles (old alignment).\n",
      " |      \n",
      " |      This controls the addition of the -profile1 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  profile2\n",
      " |      Profiles (old alignment).\n",
      " |      \n",
      " |      This controls the addition of the -profile2 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pwdnamatrix\n",
      " |      DNA weight matrix=IUB, CLUSTALW or filename\n",
      " |      \n",
      " |      This controls the addition of the -pwdnamatrix parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pwgapext\n",
      " |      Gap extension penalty\n",
      " |      \n",
      " |      This controls the addition of the -pwgapext parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pwgapopen\n",
      " |      Gap opening penalty\n",
      " |      \n",
      " |      This controls the addition of the -pwgapopen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  pwmatrix\n",
      " |      Protein weight matrix=BLOSUM, PAM, GONNET, ID or filename\n",
      " |      \n",
      " |      This controls the addition of the -pwmatrix parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  quicktree\n",
      " |      Use FAST algorithm for the alignment guide tree\n",
      " |      \n",
      " |      This property controls the addition of the -quicktree switch, treat this property as a boolean.\n",
      " |  \n",
      " |  quiet\n",
      " |      Reduce console output to minimum\n",
      " |      \n",
      " |      This property controls the addition of the -quiet switch, treat this property as a boolean.\n",
      " |  \n",
      " |  range\n",
      " |      Sequence range to write starting m to m+n. Input as string eg. '24,200'\n",
      " |      \n",
      " |      This controls the addition of the -range parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  score\n",
      " |      Either: PERCENT or ABSOLUTE\n",
      " |      \n",
      " |      This controls the addition of the -score parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  secstrout\n",
      " |      STRUCTURE or MASK or BOTH or NONE output in alignment file\n",
      " |      \n",
      " |      This controls the addition of the -secstrout parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seed\n",
      " |      Seed number for bootstraps.\n",
      " |      \n",
      " |      This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seqno_range\n",
      " |      OFF or ON (NEW- for all output formats)\n",
      " |      \n",
      " |      This controls the addition of the -seqno_range parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seqnos\n",
      " |      OFF or ON (for Clustal output only)\n",
      " |      \n",
      " |      This controls the addition of the -seqnos parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  sequences\n",
      " |      Sequentially add profile2 sequences to profile1 alignment\n",
      " |      \n",
      " |      This property controls the addition of the -sequences switch, treat this property as a boolean.\n",
      " |  \n",
      " |  stats\n",
      " |      Log some alignment statistics to file\n",
      " |      \n",
      " |      This controls the addition of the -stats parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  strandendin\n",
      " |      Number of residues inside strand to be treated as terminal\n",
      " |      \n",
      " |      This controls the addition of the -strandendin parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  strandendout\n",
      " |      Number of residues outside strand to be treated as terminal\n",
      " |      \n",
      " |      This controls the addition of the -strandendout parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  strandgap\n",
      " |      gap penalty for strand core residues\n",
      " |      \n",
      " |      This controls the addition of the -strandgap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  terminalgap\n",
      " |      Gap penalty for structure termini\n",
      " |      \n",
      " |      This controls the addition of the -terminalgap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  topdiags\n",
      " |      Number of best diags.\n",
      " |      \n",
      " |      This controls the addition of the -topdiags parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  tossgaps\n",
      " |      Ignore positions with gaps.\n",
      " |      \n",
      " |      This property controls the addition of the -tossgaps switch, treat this property as a boolean.\n",
      " |  \n",
      " |  transweight\n",
      " |      Transitions weighting\n",
      " |      \n",
      " |      This controls the addition of the -transweight parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  tree\n",
      " |      Calculate NJ tree.\n",
      " |      \n",
      " |      This property controls the addition of the -tree switch, treat this property as a boolean.\n",
      " |  \n",
      " |  type\n",
      " |      PROTEIN or DNA sequences\n",
      " |      \n",
      " |      This controls the addition of the -type parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  usetree\n",
      " |      File name of guide tree\n",
      " |      \n",
      " |      This controls the addition of the -usetree parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  usetree1\n",
      " |      File name of guide tree for profile1\n",
      " |      \n",
      " |      This controls the addition of the -usetree1 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  usetree2\n",
      " |      File name of guide tree for profile2\n",
      " |      \n",
      " |      This controls the addition of the -usetree2 parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  window\n",
      " |      Window around best diags.\n",
      " |      \n",
      " |      This controls the addition of the -window parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Execute command, wait for it to finish, return (stdout, stderr).\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output(). In general\n",
      " |      if you require more control over running the command, use subprocess\n",
      " |      directly.\n",
      " |      \n",
      " |      When the program called returns a non-zero error level, a custom\n",
      " |      ApplicationError exception is raised. This includes any stdout and\n",
      " |      stderr strings captured as attributes of the exception object, since\n",
      " |      they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  parameters = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "help(ClustalwCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "7ad53d91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustalo -infile=newcytb.fsa -outfile=newcytb.aln\n"
     ]
    }
   ],
   "source": [
    "#Step 5\n",
    "from Bio import AlignIO\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "in_file = nucleotide\n",
    "out_file = nuc_alignment\n",
    "clustalw_cline = ClustalwCommandline(\"clustalo\", infile=in_file, outfile=out_file)\n",
    "print(clustalw_cline)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "a2f5abeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Using 8 threads\\nRead 25 sequences (type: DNA) from newcytb.fsa\\nnot more sequences (25) than cluster-size (100), turn off mBed\\nCalculating pairwise ktuple-distances...\\nKtuple-distance calculation progress: 0 % (0 out of 325)\\nKtuple-distance calculation progress: 1 % (4 out of 325)\\nKtuple-distance calculation progress: 15 % (52 out of 325)\\nKtuple-distance calculation progress: 31 % (104 out of 325)\\nKtuple-distance calculation progress: 33 % (110 out of 325)\\nKtuple-distance calculation progress: 39 % (128 out of 325)\\nKtuple-distance calculation progress: 42 % (137 out of 325)\\nKtuple-distance calculation progress: 48 % (159 out of 325)\\nKtuple-distance calculation progress: 53 % (174 out of 325)\\nKtuple-distance calculation progress: 55 % (180 out of 325)\\nKtuple-distance calculation progress: 57 % (187 out of 325)\\nKtuple-distance calculation progress: 58 % (191 out of 325)\\nKtuple-distance calculation progress: 59 % (194 out of 325)\\nKtuple-distance calculation progress: 60 % (197 out of 325)\\nKtuple-distance calculation progress: 63 % (206 out of 325)\\nKtuple-distance calculation progress: 67 % (218 out of 325)\\nKtuple-distance calculation progress: 79 % (257 out of 325)\\nKtuple-distance calculation progress: 89 % (290 out of 325)\\nKtuple-distance calculation progress done. CPU time: 0.33u 0.00s 00:00:00.33 Elapsed: 00:00:00\\nGuide-tree computation done.\\nProgressive alignment progress: 4 % (1 out of 24)\\nProgressive alignment progress: 8 % (2 out of 24)\\nProgressive alignment progress: 12 % (3 out of 24)\\nProgressive alignment progress: 16 % (4 out of 24)\\nProgressive alignment progress: 20 % (5 out of 24)\\nProgressive alignment progress: 25 % (6 out of 24)\\nProgressive alignment progress: 29 % (7 out of 24)\\nProgressive alignment progress: 33 % (8 out of 24)\\nProgressive alignment progress: 37 % (9 out of 24)\\nProgressive alignment progress: 41 % (10 out of 24)\\nProgressive alignment progress: 45 % (11 out of 24)\\nProgressive alignment progress: 50 % (12 out of 24)\\nProgressive alignment progress: 54 % (13 out of 24)\\nProgressive alignment progress: 58 % (14 out of 24)\\nProgressive alignment progress: 62 % (15 out of 24)\\nProgressive alignment progress: 66 % (16 out of 24)\\nProgressive alignment progress: 70 % (17 out of 24)\\nProgressive alignment progress: 75 % (18 out of 24)\\nProgressive alignment progress: 79 % (19 out of 24)\\nProgressive alignment progress: 83 % (20 out of 24)\\nProgressive alignment progress: 87 % (21 out of 24)\\nProgressive alignment progress: 91 % (22 out of 24)\\nProgressive alignment progress: 95 % (23 out of 24)\\nProgressive alignment progress: 100 % (24 out of 24)\\nProgressive alignment progress done. CPU time: 1.48u 0.08s 00:00:01.56 Elapsed: 00:00:01\\nAlignment written to newcytb.aln\\n',\n",
       " 'WARNING: Your old-style command-line options were converted to:  clustalo -i newcytb.fsa -o newcytb.aln --outfmt=clustal -v --force\\n')"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 6\n",
    "clustalw_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "eaa9a28f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment length 1140\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAACCATCTCCT-CAAATGAAGA OP894092\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAATTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATT--AGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894093\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAATTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894094\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTACTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894095\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894096\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894097\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894098\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGGTATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACATCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894099\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTTCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGACCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894100\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894101\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894102\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAATTAGGAGGAGTCGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAATTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAATCATCTCCTCAAATGAAGA OP894103\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATTGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894104\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATTGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894105\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATTGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894106\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATTGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTTAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894107\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATTATCAATGACTCATTCATCGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCCTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACTGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATGTTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTACTATGGTTCCTACACATATACAGAAACATGAAACATTGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATCAAAGACATCCTAGGACTAGTCCTTATACTTATAGTCCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCGGCCAATCCGCTCAATACCCCACCTCATATTAAGCCAGAATGATATTTCCTATTTGCCTATGCCATCCTACGGTCAATCCCAAATAAATTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATCCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894108\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894109\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894110\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894111\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894112\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894113\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894114\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTACTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCGCTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894115\n",
      "ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAGTTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTTCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATCCTAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA OP894116\n"
     ]
    }
   ],
   "source": [
    "#Step 7\n",
    "alignment1 = AlignIO.read(open(nuc_alignment), \"clustal\")\n",
    "print(\"Alignment length %i\" % alignment1.get_alignment_length())\n",
    "for record in alignment1:\n",
    "    print(record.seq + \" \" + record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "0c4bc540",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustalo -infile=newcytb.pro -outfile=newcytb_pro.aln\n"
     ]
    }
   ],
   "source": [
    "#Step 8\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "in_file = protein_trans\n",
    "out_file = prot_alignment\n",
    "clustalw_cline = ClustalwCommandline(\"clustalo\", infile=in_file, outfile=out_file)\n",
    "print(clustalw_cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "c84387c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Using 8 threads\\nRead 25 sequences (type: Protein) from newcytb.pro\\nnot more sequences (25) than cluster-size (100), turn off mBed\\nCalculating pairwise ktuple-distances...\\nKtuple-distance calculation progress: 0 % (0 out of 325)\\nKtuple-distance calculation progress: 1 % (4 out of 325)\\nKtuple-distance calculation progress: 11 % (37 out of 325)\\nKtuple-distance calculation progress: 15 % (50 out of 325)\\nKtuple-distance calculation progress: 20 % (65 out of 325)\\nKtuple-distance calculation progress: 27 % (89 out of 325)\\nKtuple-distance calculation progress: 36 % (117 out of 325)\\nKtuple-distance calculation progress: 38 % (126 out of 325)\\nKtuple-distance calculation progress: 42 % (138 out of 325)\\nKtuple-distance calculation progress: 49 % (160 out of 325)\\nKtuple-distance calculation progress: 51 % (166 out of 325)\\nKtuple-distance calculation progress: 55 % (179 out of 325)\\nKtuple-distance calculation progress: 59 % (193 out of 325)\\nKtuple-distance calculation progress: 62 % (204 out of 325)\\nKtuple-distance calculation progress: 64 % (211 out of 325)\\nKtuple-distance calculation progress: 66 % (215 out of 325)\\nKtuple-distance calculation progress: 72 % (234 out of 325)\\nKtuple-distance calculation progress: 84 % (274 out of 325)\\nKtuple-distance calculation progress done. CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00\\nGuide-tree computation done.\\nProgressive alignment progress: 4 % (1 out of 24)\\nProgressive alignment progress: 8 % (2 out of 24)\\nProgressive alignment progress: 12 % (3 out of 24)\\nProgressive alignment progress: 16 % (4 out of 24)\\nProgressive alignment progress: 20 % (5 out of 24)\\nProgressive alignment progress: 25 % (6 out of 24)\\nProgressive alignment progress: 29 % (7 out of 24)\\nProgressive alignment progress: 33 % (8 out of 24)\\nProgressive alignment progress: 37 % (9 out of 24)\\nProgressive alignment progress: 41 % (10 out of 24)\\nProgressive alignment progress: 45 % (11 out of 24)\\nProgressive alignment progress: 50 % (12 out of 24)\\nProgressive alignment progress: 54 % (13 out of 24)\\nProgressive alignment progress: 58 % (14 out of 24)\\nProgressive alignment progress: 62 % (15 out of 24)\\nProgressive alignment progress: 66 % (16 out of 24)\\nProgressive alignment progress: 70 % (17 out of 24)\\nProgressive alignment progress: 75 % (18 out of 24)\\nProgressive alignment progress: 79 % (19 out of 24)\\nProgressive alignment progress: 83 % (20 out of 24)\\nProgressive alignment progress: 87 % (21 out of 24)\\nProgressive alignment progress: 91 % (22 out of 24)\\nProgressive alignment progress: 95 % (23 out of 24)\\nProgressive alignment progress: 100 % (24 out of 24)\\nProgressive alignment progress done. CPU time: 0.19u 0.01s 00:00:00.20 Elapsed: 00:00:00\\nAlignment written to newcytb_pro.aln\\n',\n",
       " 'WARNING: Your old-style command-line options were converted to:  clustalo -i newcytb.pro -o newcytb_pro.aln --outfmt=clustal -v --force\\n')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 9\n",
    "clustalw_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "64505119",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAETISSNE OP894092\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIISPLPTCKYR-------- OP894093\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894094\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894095\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894096\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894097\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894098\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894099\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894100\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894101\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894102\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894103\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894104\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894105\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894106\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894107\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894108\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894109\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894110\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894111\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894112\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894113\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894114\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894115\n",
      "MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW OP894116\n"
     ]
    }
   ],
   "source": [
    "#Step 10\n",
    "alignment2 = list(AlignIO.read(open(prot_alignment), \"clustal\"))\n",
    "#print(\"Alignment length %i\" % alignment2.get_alignment_length())\n",
    "for record in alignment2:\n",
    "    print(record.seq + \" \" + record.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8007060",
   "metadata": {},
   "source": [
    "## Convert alignments from clustal to fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "c2be75f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 11\n",
    "from Bio import AlignIO\n",
    "alignment3 = AlignIO.read(nuc_alignment, \"clustal\")\n",
    "nuc_alignment_fsa  = SeqIO.write(alignment3, \"newcytb.fsaaln\", \"fasta\")\n",
    "alignment4 = AlignIO.read(prot_alignment, \"clustal\")\n",
    "prot_alignment_fsa  = SeqIO.write(alignment4, \"newcytb_pro.fsaaln\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03bc55dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c4fc372d",
   "metadata": {},
   "source": [
    "## Score alignments\n",
    "### https://github.com/benhid/pyMSA/blob/master/resources/tutorial-pymsa.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f2cfeb50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyMSA\n",
      "  Downloading pyMSA-0.8.1.tar.gz (10 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hBuilding wheels for collected packages: pyMSA\n",
      "  Building wheel for pyMSA (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pyMSA: filename=pyMSA-0.8.1-py3-none-any.whl size=12824 sha256=a71b85fe50802981c74b4b334beb49f7000a39468af1baa4e08c2d21e2e21e05\n",
      "  Stored in directory: /Users/rich/Library/Caches/pip/wheels/c7/00/54/1acf56628af0e3684b5ba4395448fa3ca59700ce22994e5935\n",
      "Successfully built pyMSA\n",
      "Installing collected packages: pyMSA\n",
      "Successfully installed pyMSA-0.8.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pyMSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "b61a144d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 12\n",
    "#file_name = nuc_alignment_fsa \n",
    "list_of_pairs = []\n",
    "key = ''\n",
    "value = ''\n",
    "\n",
    "with open(\"newcytb_pro.fsaaln\", 'r') as file:\n",
    "    for line in file:\n",
    "        if line[0] == '>':\n",
    "            if key != '':\n",
    "                list_of_pairs.append((key, value))\n",
    "            key = line[1:].rstrip()\n",
    "            value = ''\n",
    "        else:\n",
    "            value += line.rstrip()\n",
    "\n",
    "list_of_pairs.append((key, value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2d40251d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(list_of_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "55ff2862",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 12 alternate\n",
    "from pymsa.core.msa import MSA\n",
    "\n",
    "\n",
    "def read_fasta_file_as_list_of_pairs(file_name: str) -> list:\n",
    "    \"\"\"\n",
    "    Read a file in FASTA format as list of pairs (sequence id, sequence).\n",
    "\n",
    "    :param file_name: FASTA file.\n",
    "    :return: List of pairs.\n",
    "    \"\"\"\n",
    "    list_of_pairs = []\n",
    "    key = ''\n",
    "    value = ''\n",
    "\n",
    "    with open(file_name, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[0] == '>':\n",
    "                if key != '':\n",
    "                    list_of_pairs.append((key, value))\n",
    "                key = line[1:].rstrip()\n",
    "                value = ''\n",
    "            else:\n",
    "                value += line.rstrip()\n",
    "\n",
    "    list_of_pairs.append((key, value))\n",
    "    return list_of_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b457a1b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('OP894092',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAETISSNE'),\n",
       " ('OP894093',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIISPLPTCKYR--------'),\n",
       " ('OP894094',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894095',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894096',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894097',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894098',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894099',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894100',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894101',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894102',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894103',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894104',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894105',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894106',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894107',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894108',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894109',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894110',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894111',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894112',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894113',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894114',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894115',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894116',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW')]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 12 alternate\n",
    "file_name = \"newcytb_pro.fsaaln\"\n",
    "read_fasta_file_as_list_of_pairs(file_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "7e3c1b2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentage of non-gaps: 99.9155672823219 %\n",
      "Percentage of totally conserved columns: 95.51451187335093\n",
      "Entropy score: -3.9635377263906153\n",
      "Sum of Pairs score (Blosum62): 597014\n",
      "Sum of Pairs score (PAM250): 617661\n",
      "Sum of Pairs score (PAM380): 656838\n",
      "Star score (Blosum62): 49961\n",
      "Star score (PAM250): 51696\n"
     ]
    }
   ],
   "source": [
    "#Step 13\n",
    "from pymsa import MSA, Entropy, PercentageOfNonGaps, PercentageOfTotallyConservedColumns, Star, SumOfPairs\n",
    "from pymsa import PAM250, Blosum62, FileMatrix\n",
    "from pymsa.util.fasta import print_alignment\n",
    "\n",
    "\n",
    "def run_all_scores(sequences: list) -> None:\n",
    "    aligned_sequences = list(pair[1] for pair in sequences)\n",
    "    sequences_id = list(pair[0] for pair in sequences)\n",
    "\n",
    "    msa = MSA(aligned_sequences, sequences_id)\n",
    "    #print_alignment(msa)\n",
    "\n",
    "    # Percentage of non-gaps and totally conserved columns\n",
    "    non_gaps = PercentageOfNonGaps(msa)\n",
    "    totally_conserved_columns = PercentageOfTotallyConservedColumns(msa)\n",
    "\n",
    "    percentage = non_gaps.compute()\n",
    "    print(\"Percentage of non-gaps: {0} %\".format(percentage))\n",
    "\n",
    "    conserved = totally_conserved_columns.compute()\n",
    "    print(\"Percentage of totally conserved columns: {0}\".format(conserved))\n",
    "\n",
    "    # Entropy\n",
    "    value = Entropy(msa).compute()\n",
    "    print(\"Entropy score: {0}\".format(value))\n",
    "\n",
    "    # Sum of pairs\n",
    "    value = SumOfPairs(msa, Blosum62()).compute()\n",
    "    print(\"Sum of Pairs score (Blosum62): {0}\".format(value))\n",
    "\n",
    "    value = SumOfPairs(msa, PAM250()).compute()\n",
    "    print(\"Sum of Pairs score (PAM250): {0}\".format(value))\n",
    "\n",
    "    value = SumOfPairs(msa, FileMatrix('PAM380.txt')).compute()\n",
    "    print(\"Sum of Pairs score (PAM380): {0}\".format(value))\n",
    "\n",
    "    # Star\n",
    "    value = Star(msa, Blosum62()).compute()\n",
    "    print(\"Star score (Blosum62): {0}\".format(value))\n",
    "\n",
    "    value = Star(msa, PAM250()).compute()\n",
    "    print(\"Star score (PAM250): {0}\".format(value))\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sequences = [(\"1g41\",\n",
    "                  \"S-EMTPREIVSELDQHIIGQADAKRAVAIALRNRWRRMQLQEPLRHE--------VTP-KNILMIGPTGVGKTEIARRLAKLANAPFIKVEATKFT----\"\n",
    "                  \"VGKEVDSIIRDLTDSAMKLVRQQEIAKNR---------------------------------------------------------------------LI\"\n",
    "                  \"DDEAAKLINPEELKQKAIDAVE--QNGIVFIDEIDKICKKGEYSGADVSREGVQRDLLPLVEGSTVSTKHGMVKTDHILFIASGAFQVARPSDL------\"\n",
    "                  \"-----------IPELQGRLPIR-VEL---TALSAADFERILTEPHASLTEQYKALMATEGVNIAFTTDAVKKIAEAAFRVNEKTENIGARRLHTVMERLM\"\n",
    "                  \"DKISFSASDMNGQTVNIDAAYVADALGEVVENEDLSRFIL\"),\n",
    "                 (\"1e94\",\n",
    "                  \"HSEMTPREIVSELDKHIIGQDNAKRSVAIALRNRWRRMQLNEELRHE--------VTP-KNILMIGPTGVGKTEIARRLAKLANAPFIKVEATKFTEVGY\"\n",
    "                  \"VGKEVDSIIRDLTDAAVKMVRVQAIEKNRYRAEELAEERILDVLIPPAKNNWGQTEQQQEPSAARQAFRKKLREGQLDDKEIEKQKARKLKIKDAMKLLI\"\n",
    "                  \"EEEAAKLVNPEELKQDAIDAVE--QHGIVFIDEIDKICKRGESSGPDVSREGVQRDLLPLVEGCTVSTKHGMVKTDHILFIASGAFQIAKPSDL------\"\n",
    "                  \"-----------IPELQGRLPIR-VEL---QALTTSDFERILTEPNASITVQYKALMATEGVNIEFTDSGIKRIAEAAWQVNESTENIGARRLHTVLERLM\"\n",
    "                  \"EEISYDASDLSGQNITIDADYVSKHLDALVADEDLSRFIL\"),\n",
    "                 (\"1e32\",\n",
    "                  \"R-ED-EEESLNEVGYDDVGG--CRKQLAQ-----I-KEMVELPLRHPALFKAIGVKPP-RGILLYGPPGTGKTLIARAVANETGAFFFLINGPEIM-SKL\"\n",
    "                  \"A-GESESN--------------------------------------------------------------------------------------------\"\n",
    "                  \"-------------LRKAFEEAEKNAPAIIFIDELDAIAPKREKTHGEVERRIVSQ-LLTLMDGL--------KQRAHVIVMAATN----RPNSIDPALRR\"\n",
    "                  \"FGRFDREVDIGIPDATGRLEILQIHTKNMKLADDVDLEQVANETHGH---------------------------------------VGADLAALCSEAAL\"\n",
    "                  \"QAIRKKMDLIDLEDETIDAEVM-NSL-AVTMDDFRWALSQ\"),\n",
    "                 (\"1d2n\",\n",
    "                  \"------EDYASYIMNGIIKWGDP---VTRVLD--DGELLVQQTKNSD--------RTPLVSVLLEGPPHSGKTALAAKIAEESNFPFIKICSPDKM-IGF\"\n",
    "                  \"SETAKCQA--------------------------------------------------------------------------------------------\"\n",
    "                  \"-------------MKKIFDDAYKSQLSCVVVDDIERLLDYV-PIGPRFSNLVLQA-LLVLLKKA-------PPQGRKLLIIGTTS----R-KDVLQEMEM\"\n",
    "                  \"LNA---------------------------------FSTTIHVPNIATGEQL--LEALEL-LGNFKDKE---RTTIAQQVKGKKVWIGIKKLLMLIEM--\"\n",
    "                  \"-------------SLQMDPEYRVRKFLALLREEGAS-PLD\")]\n",
    "    #run_all_scores(sequences)\n",
    "    run_all_scores(list_of_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e48b65c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81fc64a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "7035df1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('OP894092',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAETISSNE'),\n",
       " ('OP894093',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIISPLPTCKYR--------'),\n",
       " ('OP894094',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894095',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894096',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894097',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894098',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894099',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894100',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894101',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894102',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894103',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894104',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894105',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894106',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894107',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894108',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894109',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894110',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894111',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894112',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894113',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894114',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894115',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894116',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW')]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Skip\n",
    "#print(file_name)\n",
    "read_fasta_file_as_list_of_pairs(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "6410340a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "eecf0efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This didn't work but may not need it\n",
    "#from Bio.Align import AlignInfo\n",
    "#summary_align = AlignInfo.SummaryInfo(alignment)\n",
    "#consensus = summary_align.replacement_dictionary()\n",
    "#print(consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ea10f38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "a3d8b7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This wasn't that informative but save for now\n",
    "#substitutions = alignment2.substitutions\n",
    "#print(substitutions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "430c9e30",
   "metadata": {},
   "source": [
    "### Iterate through alignments and finds sequences containing single or double gaps that are reported to the user. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "b3c0248e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "single gap found in OP894092 at location 1131\n",
      "double gap found in OP894093 at location 1089\n"
     ]
    }
   ],
   "source": [
    "#Step 14\n",
    "single_gap_id = []\n",
    "double_gap_id = []\n",
    "for record in alignment1:\n",
    "#for record in nuc_align_fsa:\n",
    "    loc = []\n",
    "    seqid = record.id\n",
    "    sequence = str(record.seq)\n",
    "    #regex_single = r'[A-Z]+-[A-Z]'\n",
    "    #regex_double = r'[A-Z]+--[A-Z]'\n",
    "    regex_single = re.compile(r'[A-Z]+-[A-Z]')\n",
    "    regex_double = re.compile(r'[A-Z]+--[A-Z]')\n",
    "    \n",
    "    single_gap = re.search(regex_single, sequence)\n",
    "    if single_gap is not None:\n",
    "        loc = single_gap.end()\n",
    "        print('single gap found in', record.id, 'at location', loc)\n",
    "        single_gap_id.append(seqid)\n",
    "    \n",
    "    double_gap = re.search(regex_double, sequence)\n",
    "    if double_gap is not None:\n",
    "        loc = double_gap.end()\n",
    "        print('double gap found in', record.id, 'at location', loc)\n",
    "        double_gap_id.append(seqid)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ccf4d1c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "486badbc",
   "metadata": {},
   "source": [
    "### Replace gap characters with Ns, retranslate and save new nucleotide and protein files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "96f078e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 15\n",
    "#This works and does everything as a seq_record within biopython\n",
    "from Bio.Seq import MutableSeq\n",
    "sequences = []\n",
    "for index, seq_record in enumerate(SeqIO.parse(\"newcytb.fsaaln\", \"fasta\")):\n",
    "    s = seq_record\n",
    "    if s.id in single_gap_id:\n",
    "        sequence = MutableSeq(s.seq)\n",
    "        location = sequence.index(\"-\")\n",
    "        sequence[location] = 'N'\n",
    "        s.seq = sequence\n",
    "        sequences.append(s)\n",
    "        #print(seq_record.id, seq_record.seq)\n",
    "    \n",
    "    elif s.id in double_gap_id:\n",
    "        sequence = MutableSeq(s.seq)  \n",
    "        location = sequence.index(\"-\")\n",
    "        sequence[location] = 'N'\n",
    "        sequence[location + 1] = 'N'\n",
    "        s.seq = sequence\n",
    "        sequences.append(s)\n",
    "        #print(seq_record.id, seq_record.seq)\n",
    "    else:\n",
    "        sequences.append(s)\n",
    "SeqIO.write(sequences, \"revisedfasta\", \"fasta\")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "430a690b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 16\n",
    "protein_seq = []\n",
    "for seq_record in SeqIO.parse(\"revisedfasta\", \"fasta\"):\n",
    "    s = seq_record\n",
    "    s.seq = s.seq.translate(table = translation_table, to_stop=True)\n",
    "    protein_seq.append(s)\n",
    "SeqIO.write(protein_seq, revised_pro, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63080338",
   "metadata": {},
   "source": [
    "## Repeat alignments and score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "8186e63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustalo -infile=revisedfasta -outfile=revisedfasta.aln\n"
     ]
    }
   ],
   "source": [
    "#Step 17 nucleotide\n",
    "in_file = revised_nucs\n",
    "out_file = revised_nucs_aln\n",
    "clustalw_cline = ClustalwCommandline(\"clustalo\", infile=in_file, outfile=out_file)\n",
    "print(clustalw_cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "d1119ff6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Using 8 threads\\nRead 25 sequences (type: DNA) from revisedfasta\\nnot more sequences (25) than cluster-size (100), turn off mBed\\nCalculating pairwise ktuple-distances...\\nKtuple-distance calculation progress: 0 % (0 out of 325)\\nKtuple-distance calculation progress: 1 % (4 out of 325)\\nKtuple-distance calculation progress: 20 % (68 out of 325)\\nKtuple-distance calculation progress: 33 % (110 out of 325)\\nKtuple-distance calculation progress: 34 % (111 out of 325)\\nKtuple-distance calculation progress: 42 % (137 out of 325)\\nKtuple-distance calculation progress: 44 % (144 out of 325)\\nKtuple-distance calculation progress: 48 % (157 out of 325)\\nKtuple-distance calculation progress: 51 % (167 out of 325)\\nKtuple-distance calculation progress: 57 % (186 out of 325)\\nKtuple-distance calculation progress: 62 % (202 out of 325)\\nKtuple-distance calculation progress: 63 % (207 out of 325)\\nKtuple-distance calculation progress: 64 % (211 out of 325)\\nKtuple-distance calculation progress: 67 % (220 out of 325)\\nKtuple-distance calculation progress: 69 % (227 out of 325)\\nKtuple-distance calculation progress: 70 % (230 out of 325)\\nKtuple-distance calculation progress: 76 % (248 out of 325)\\nKtuple-distance calculation progress: 82 % (268 out of 325)\\nKtuple-distance calculation progress: 89 % (291 out of 325)\\nKtuple-distance calculation progress done. CPU time: 0.34u 0.00s 00:00:00.34 Elapsed: 00:00:00\\nGuide-tree computation done.\\nProgressive alignment progress: 4 % (1 out of 24)\\nProgressive alignment progress: 8 % (2 out of 24)\\nProgressive alignment progress: 12 % (3 out of 24)\\nProgressive alignment progress: 16 % (4 out of 24)\\nProgressive alignment progress: 20 % (5 out of 24)\\nProgressive alignment progress: 25 % (6 out of 24)\\nProgressive alignment progress: 29 % (7 out of 24)\\nProgressive alignment progress: 33 % (8 out of 24)\\nProgressive alignment progress: 37 % (9 out of 24)\\nProgressive alignment progress: 41 % (10 out of 24)\\nProgressive alignment progress: 45 % (11 out of 24)\\nProgressive alignment progress: 50 % (12 out of 24)\\nProgressive alignment progress: 54 % (13 out of 24)\\nProgressive alignment progress: 58 % (14 out of 24)\\nProgressive alignment progress: 62 % (15 out of 24)\\nProgressive alignment progress: 66 % (16 out of 24)\\nProgressive alignment progress: 70 % (17 out of 24)\\nProgressive alignment progress: 75 % (18 out of 24)\\nProgressive alignment progress: 79 % (19 out of 24)\\nProgressive alignment progress: 83 % (20 out of 24)\\nProgressive alignment progress: 87 % (21 out of 24)\\nProgressive alignment progress: 91 % (22 out of 24)\\nProgressive alignment progress: 95 % (23 out of 24)\\nProgressive alignment progress: 100 % (24 out of 24)\\nProgressive alignment progress done. CPU time: 1.46u 0.07s 00:00:01.53 Elapsed: 00:00:01\\nAlignment written to revisedfasta.aln\\n',\n",
       " 'WARNING: Your old-style command-line options were converted to:  clustalo -i revisedfasta -o revisedfasta.aln --outfmt=clustal -v --force\\n')"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 18\n",
    "clustalw_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "4c06e016",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustalo -infile=revisedprotein -outfile=revisedprotein.aln\n"
     ]
    }
   ],
   "source": [
    "#Step 18 protein\n",
    "in_file = revised_pro\n",
    "out_file = revised_pro_aln\n",
    "clustalw_cline = ClustalwCommandline(\"clustalo\", infile=revised_pro, outfile=revised_pro_aln)\n",
    "print(clustalw_cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "dfc2918d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Using 8 threads\\nRead 25 sequences (type: Protein) from revisedprotein\\nnot more sequences (25) than cluster-size (100), turn off mBed\\nCalculating pairwise ktuple-distances...\\nKtuple-distance calculation progress: 0 % (0 out of 325)\\nKtuple-distance calculation progress: 1 % (6 out of 325)\\nKtuple-distance calculation progress: 2 % (7 out of 325)\\nKtuple-distance calculation progress: 11 % (39 out of 325)\\nKtuple-distance calculation progress: 12 % (40 out of 325)\\nKtuple-distance calculation progress: 17 % (57 out of 325)\\nKtuple-distance calculation progress: 23 % (75 out of 325)\\nKtuple-distance calculation progress: 25 % (82 out of 325)\\nKtuple-distance calculation progress: 29 % (97 out of 325)\\nKtuple-distance calculation progress: 33 % (109 out of 325)\\nKtuple-distance calculation progress: 36 % (117 out of 325)\\nKtuple-distance calculation progress: 40 % (131 out of 325)\\nKtuple-distance calculation progress: 43 % (143 out of 325)\\nKtuple-distance calculation progress: 45 % (147 out of 325)\\nKtuple-distance calculation progress: 46 % (151 out of 325)\\nKtuple-distance calculation progress: 47 % (153 out of 325)\\nKtuple-distance calculation progress: 52 % (172 out of 325)\\nKtuple-distance calculation progress: 62 % (202 out of 325)\\nKtuple-distance calculation progress: 63 % (205 out of 325)\\nKtuple-distance calculation progress: 85 % (277 out of 325)\\nKtuple-distance calculation progress done. CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00\\nGuide-tree computation done.\\nProgressive alignment progress: 4 % (1 out of 24)\\nProgressive alignment progress: 8 % (2 out of 24)\\nProgressive alignment progress: 12 % (3 out of 24)\\nProgressive alignment progress: 16 % (4 out of 24)\\nProgressive alignment progress: 20 % (5 out of 24)\\nProgressive alignment progress: 25 % (6 out of 24)\\nProgressive alignment progress: 29 % (7 out of 24)\\nProgressive alignment progress: 33 % (8 out of 24)\\nProgressive alignment progress: 37 % (9 out of 24)\\nProgressive alignment progress: 41 % (10 out of 24)\\nProgressive alignment progress: 45 % (11 out of 24)\\nProgressive alignment progress: 50 % (12 out of 24)\\nProgressive alignment progress: 54 % (13 out of 24)\\nProgressive alignment progress: 58 % (14 out of 24)\\nProgressive alignment progress: 62 % (15 out of 24)\\nProgressive alignment progress: 66 % (16 out of 24)\\nProgressive alignment progress: 70 % (17 out of 24)\\nProgressive alignment progress: 75 % (18 out of 24)\\nProgressive alignment progress: 79 % (19 out of 24)\\nProgressive alignment progress: 83 % (20 out of 24)\\nProgressive alignment progress: 87 % (21 out of 24)\\nProgressive alignment progress: 91 % (22 out of 24)\\nProgressive alignment progress: 95 % (23 out of 24)\\nProgressive alignment progress: 100 % (24 out of 24)\\nProgressive alignment progress done. CPU time: 0.18u 0.01s 00:00:00.19 Elapsed: 00:00:00\\nAlignment written to revisedprotein.aln\\n',\n",
       " 'WARNING: Your old-style command-line options were converted to:  clustalo -i revisedprotein -o revisedprotein.aln --outfmt=clustal -v --force\\n')"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 19\n",
    "clustalw_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "7ce046c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 20\n",
    "alignment4 = AlignIO.read(\"revised_protein_trans.aln\", \"clustal\")\n",
    "SeqIO.write(alignment4, \"revised_protein_trans.alnfsa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "72303e9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('OP894092',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAETISXKW'),\n",
       " ('OP894093',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIIXVLFPLASIAENHLLKW'),\n",
       " ('OP894094',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWILRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894095',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894096',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894097',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894098',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894099',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894100',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894101',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894102',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894103',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894104',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894105',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894106',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894107',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894108',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894109',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894110',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894111',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894112',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894113',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894114',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894115',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW'),\n",
       " ('OP894116',\n",
       "  'MTNIRKSHPLFKIINDSFIDLPAPSSISSWWNFGSLLGVCLAIQILTGLFLAMHYTSDTATAFYSVTHICRDVNYGWVLRYLHANGASMFFICLFLHVGRGIYYGSYTYTETWNIGIILLFAVMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTNLVEWVWGGFSVDKATLTRFFALHFLLPFIIAALVMVHLLFLHETGSNNPTGMPSDVDMIPFHPYYTIKDILGLVLMLMVLLSLVLFSPDLLGDPDNYTPANPLNTPPHIKPEWYFLFAYAILRSIPNKLGGVVALVLSILILAVIPLLHTSKQRSMTFRPISQCLFWLLVADLLTLTWIGGQPVEHPFILIGQLASILYFLIILVLFPLASIAENHLLKW')]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 21\n",
    "file_name = \"revised_protein_trans.alnfsa\"\n",
    "read_fasta_file_as_list_of_pairs(file_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "363c9a2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentage of non-gaps: 99.9155672823219 %\n",
      "Percentage of totally conserved columns: 95.51451187335093\n",
      "Entropy score: -3.9635377263906153\n",
      "Sum of Pairs score (Blosum62): 597014\n",
      "Sum of Pairs score (PAM250): 617661\n",
      "Sum of Pairs score (PAM380): 656838\n",
      "Star score (Blosum62): 49961\n",
      "Star score (PAM250): 51696\n"
     ]
    }
   ],
   "source": [
    "#Step 22\n",
    "run_all_scores(list_of_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b5c7769",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "df3c003a",
   "metadata": {},
   "source": [
    "### Test blocks and extra code not being used. Saving for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "c04a801b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1086\n",
      "OP894093 ATGACCAACATCCGAAAATCCCACCCACTATTCAAAATCATCAATGACTCATTTATTGACCTTCCCGCCCCCTCAAGTATCTCATCCTGATGAAATTTTGGCTCCCTATTAGGCGTATGCTTAGCAATCCAAATCCTAACAGGCCTATTCCTAGCAATACATTACACATCTGACACCGCTACCGCCTTCTACTCCGTCACCCATATCTGCCGAGACGTCAACTACGGCTGAATTCTGCGCTATCTCCACGCCAACGGAGCATCCATATTCTTCATCTGCCTATTCCTACATGTAGGCCGAGGCATCTATTATGGTTCCTACACATATACAGAAACATGAAACATCGGAATCATCCTTCTCTTTGCCGTTATAGCAACAGCCTTCATAGGATATGTCCTCCCATGAGGACAAATATCTTTCTGAGGCGCAACAGTCATCACCAATCTCCTCTCAGCTATCCCCTATATCGGAACTAATCTCGTAGAGTGAGTTTGAGGGGGCTTCTCCGTAGACAAAGCTACCCTCACTCGATTCTTCGCCCTCCACTTTCTCCTCCCATTTATTATCGCAGCCCTAGTAATAGTACATCTGCTCTTCCTACATGAAACAGGCTCAAACAACCCCACAGGAATACCATCAGATGTAGACATAATCCCCTTCCACCCTTACTACACAATTAAAGATATTCTAGGACTAGTCCTCATACTTATAGTTCTATTATCACTAGTGCTATTCTCACCAGACCTCCTAGGAGACCCAGACAACTACACCCCAGCCAACCCGCTCAACACCCCACCCCATATTAAGCCAGAATGATACTTCCTATTTGCCTATGCCATCCTACGATCAATCCCAAATAAACTAGGAGGAGTTGTAGCCCTAGTCCTTTCCATTCTTATCCTAGCCGTAATTCCCCTACTTCACACATCAAAACAACGCAGCATAACCTTCCGACCTATCAGCCAATGCCTATTCTGACTTCTAGTTGCTGATCTCCTCACACTAACATGAATTGGCGGACAACCTGTAGAACACCCATTCATCCTCATTGGACAACTAGCCTCAATCCTCTACTTCCTTATTATTNNAGTCCTCTTCCCACTTGCAAGTATCGCAGAAAACCATCTCCTCAAATGAAGA\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#testing options\n",
    "from Bio.Seq import MutableSeq\n",
    "sequences = []\n",
    "for index, seq_record in enumerate(SeqIO.parse(nuc_align_fsa, \"fasta\")):\n",
    "    for id in double_gap_id:\n",
    "        if id in seq_record.id:\n",
    "            sequence = MutableSeq(seq_record.seq)\n",
    "            location = sequence.index(\"-\")\n",
    "            print(location)\n",
    "            sequence[location] = 'N'\n",
    "            sequence[location + 1] = 'N'\n",
    "            #regex_double_replace = re.compile(\"[--]\")\n",
    "            #sequence = regex_double_replace.sub(\"N\", sequence)\n",
    "            seq_record.seq = sequence\n",
    "            print(seq_record.id, seq_record.seq)\n",
    "SeqIO.write(seq_record, \"revisedfasta\", \"fasta\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9152f57f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Skip I moved it up\n",
    "from pymsa.core.msa import MSA\n",
    "\n",
    "\n",
    "def read_fasta_file_as_list_of_pairs(file_name: str) -> list:\n",
    "    \"\"\"\n",
    "    Read a file in FASTA format as list of pairs (sequence id, sequence).\n",
    "\n",
    "    :param file_name: FASTA file.\n",
    "    :return: List of pairs.\n",
    "    \"\"\"\n",
    "    list_of_pairs = []\n",
    "    key = ''\n",
    "    value = ''\n",
    "\n",
    "    with open(file_name, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[0] == '>':\n",
    "                if key != '':\n",
    "                    list_of_pairs.append((key, value))\n",
    "                key = line[1:].rstrip()\n",
    "                value = ''\n",
    "            else:\n",
    "                value += line.rstrip()\n",
    "\n",
    "    list_of_pairs.append((key, value))\n",
    "    return list_of_pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2317eebe",
   "metadata": {},
   "source": [
    "### Iterates though alignments, finds gaps, prints list of seqids and replaces single or double gap with N. Repeat of above but may not be the best solution REQUIRES STRING OUTPUT SKIPPING FOR NOW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ec290a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_gap_id = []\n",
    "double_gap_id = []\n",
    "for record in alignment1:\n",
    "    seqid = record.id\n",
    "    sequence = str(record.seq)\n",
    "    regex_single = r'[A-Z]+-[A-Z]'\n",
    "    regex_double = r'[A-Z]+--[A-Z]'\n",
    "    regex_single_replace = re.compile(\"[-]\")\n",
    "    regex_double_replace = re.compile(\"[--]\")\n",
    "    \n",
    "    single_gap = re.search(regex_single, sequence)\n",
    "    if single_gap:\n",
    "        print('single gap found', record.id)\n",
    "        single_gap_id.append(seqid)\n",
    "        sequence = regex_single_replace.sub(\"N\", sequence)\n",
    "        record.seq = sequence\n",
    "        print(record.id, record.seq)\n",
    "        \n",
    "    \n",
    "    double_gap = re.search(regex_double, sequence)\n",
    "    if double_gap:\n",
    "        print('double gap found', record.id)\n",
    "        double_gap_id.append(seqid)\n",
    "        sequence = regex_double_replace.sub(\"NN\", sequence)\n",
    "        record.seq = sequence\n",
    "        print(record.id, record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22ac3e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_seq = []\n",
    "for seq_record in SeqIO.parse(\"revisedfasta\", \"fasta\"):\n",
    "    s = seq_record\n",
    "    s.seq = s.seq.translate(table = translation_table, to_stop=True)\n",
    "    protein_seq.append(s)\n",
    "#print(protein_seq)\n",
    "SeqIO.write(protein_seq, \"revised_protein_trans\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed9023b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This works but as string and would require the output to be written as a string. SeqIO.write will not work here\n",
    "#Save for now may be useful if want to do global replacements\n",
    "sequences = []\n",
    "for index, seq_record in enumerate(SeqIO.parse(nuc_alignment_fsa, \"fasta\")):\n",
    "    s = seq_record\n",
    "    if seq_record.id in single_gap_id:\n",
    "        sequence = str(seq_record.seq) \n",
    "        regex_single_replace = re.compile(\"[-]\")\n",
    "        sequence = regex_double_replace.sub(\"N\", sequence)\n",
    "        seq_record.seq = sequence\n",
    "        sequences.append(s)\n",
    "        print(seq_record.id, seq_record.seq)\n",
    "    \n",
    "    if seq_record.id in double_gap_id:\n",
    "        sequence = str(seq_record.seq)   \n",
    "        regex_double_replace = re.compile(\"[--]\")\n",
    "        sequence = regex_double_replace.sub(\"N\", sequence)\n",
    "        seq_record.seq = sequence\n",
    "        sequences.append(s)\n",
    "    else:\n",
    "        sequences.append(s)\n",
    "       \n",
    "#SeqIO.write(sequences, \"revisedfasta\", \"fasta\")  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
